hypothesis-python/src/hypothesis/internal/conjecture/engine.py

```diff
@@ -199,6 +206,8 @@ def test_function(self, data):
                 self.exit_with(ExitReason.max_shrinks)
 
         if not self.interesting_examples:
+            # Note that this logic is reproduced to end the generation phase when we
```

Comment:
Why is the logic reproduced as upposed to just have `generate_new_examples` be a method? (I think I remember there being a reason, but I don't remember what it is, which suggests that if this comment is going to be here it should probably explain the reasoning)

Reply:
It's because in `test_function` (here), we exit the engine with some particular cause if we have reached  the thresholds for valid_examples or call_count without finding any bugs.  In the latter case, we just want to end the generation phase when, after finding a first bug, we reach a threshold that would have us stop looking if we had not found one.

They're only related by mimicry rather than principle, and guard different behaviour, so IMO leaving an explanatory comment is the least clumsy way of linking them.  I've expanded the comments at each end to explain this logic though :+1: